File: src/test/java/com/zepben/vertxutils/routing/RoutingContextExTest.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing;

import com.zepben.vertxutils.routing.handlers.params.BadParamException;
import com.zepben.vertxutils.routing.handlers.params.PathParams;
import com.zepben.vertxutils.routing.handlers.params.QueryParams;
import io.vertx.ext.web.RoutingContext;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.Objects;

import static com.zepben.testutils.exception.ExpectException.expect;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.mockito.Mockito.*;

public class RoutingContextExTest {

    private final RoutingContext context = mock(RoutingContext.class);

    @Test
    public void keysAreDifferent() {
        assertThat(RoutingContextEx.INSTANCE.getPATH_PARAMS_KEY(), not(equalTo(RoutingContextEx.INSTANCE.getQUERY_PARAMS_KEY())));
        assertThat(RoutingContextEx.INSTANCE.getPATH_PARAMS_KEY(), not(equalTo(RoutingContextEx.INSTANCE.getBODY_KEY())));
    }

    @Test
    public void getPathParams() {
        PathParams params = new PathParams(Collections.emptyMap());
        doReturn(params).when(context).get(RoutingContextEx.INSTANCE.getPATH_PARAMS_KEY());
        assertThat(RoutingContextEx.INSTANCE.getPathParams(context), is(params));
    }

    @Test
    public void getPathParamsWhenNoParams() {
        expect(() -> RoutingContextEx.INSTANCE.getPathParams(context)).toThrow(IllegalStateException.class);
    }

    @Test
    public void putPathParams() {
        PathParams params = new PathParams(Collections.emptyMap());
        RoutingContextEx.INSTANCE.putPathParams(context, params);
        verify(context).put(RoutingContextEx.INSTANCE.getPATH_PARAMS_KEY(), params);
    }

    @Test
    public void getQueryParams() {
        QueryParams params = mock(QueryParams.class);
        doReturn(params).when(context).get(RoutingContextEx.INSTANCE.getQUERY_PARAMS_KEY());
        assertThat(RoutingContextEx.INSTANCE.getQueryParams(context), is(params));
    }

    @Test
    public void getQueryParamsWhenNoParams() {
        expect(() -> RoutingContextEx.INSTANCE.getQueryParams(context)).toThrow(IllegalStateException.class);
    }

    @Test
    public void putQueryParams() {
        QueryParams params = mock(QueryParams.class);
        RoutingContextEx.INSTANCE.putQueryParams(context, params);
        verify(context).put(RoutingContextEx.INSTANCE.getQUERY_PARAMS_KEY(), params);
    }

    @Test
    public void getDecodedBody() {
        doReturn("expected").when(context).get(RoutingContextEx.INSTANCE.getBODY_KEY());
        assertThat(RoutingContextEx.INSTANCE.getDecodedBody(context), is("expected"));
    }

    @Test
    public void getOptionalDecodedBody() {
        doReturn("expected").when(context).get(RoutingContextEx.INSTANCE.getBODY_KEY());
        assertThat(Objects.requireNonNull(RoutingContextEx.INSTANCE.getOptionalDecodedBody(context)), is("expected"));
    }

    @Test
    public void getMissingOptionalDecodedBody() {
        assertThat(RoutingContextEx.INSTANCE.getOptionalDecodedBody(context), nullValue());
    }

    @Test
    public void geRequestBodyWhenNoBody() {
        expect(() -> RoutingContextEx.INSTANCE.getDecodedBody(context)).toThrow(BadParamException.class);
    }

    @Test
    public void putRequestBody() {
        Object body = new Object();
        RoutingContextEx.INSTANCE.putRequestBody(context, body);
        verify(context).put(RoutingContextEx.INSTANCE.getBODY_KEY(), body);
    }

}

File: src/test/java/com/zepben/vertxutils/routing/RoutingContextExTest.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.routing

import com.zepben.testutils.exception.ExpectException.Companion.expect
import com.zepben.testutils.junit.SystemLogExtension
import com.zepben.vertxutils.routing.RoutingContextEx.BODY_KEY
import com.zepben.vertxutils.routing.RoutingContextEx.PATH_PARAMS_KEY
import com.zepben.vertxutils.routing.RoutingContextEx.QUERY_PARAMS_KEY
import com.zepben.vertxutils.routing.RoutingContextEx.getDecodedBody
import com.zepben.vertxutils.routing.RoutingContextEx.getOptionalDecodedBody
import com.zepben.vertxutils.routing.RoutingContextEx.getPathParams
import com.zepben.vertxutils.routing.RoutingContextEx.getQueryParams
import com.zepben.vertxutils.routing.RoutingContextEx.putPathParams
import com.zepben.vertxutils.routing.RoutingContextEx.putQueryParams
import com.zepben.vertxutils.routing.RoutingContextEx.putRequestBody
import com.zepben.vertxutils.routing.handlers.params.BadParamException
import com.zepben.vertxutils.routing.handlers.params.PathParams
import com.zepben.vertxutils.routing.handlers.params.QueryParams
import io.vertx.ext.web.RoutingContext
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.RegisterExtension
import org.mockito.Mockito.*

class RoutingContextExTest {

    companion object {

        @JvmField
        @RegisterExtension
        val systemOut: SystemLogExtension = SystemLogExtension.SYSTEM_OUT.captureLog().muteOnSuccess()

    }

    private val context = mock<RoutingContext>()

    @Test
    fun keysAreDifferent() {
        assertThat(PATH_PARAMS_KEY, not(equalTo(QUERY_PARAMS_KEY)))
        assertThat(PATH_PARAMS_KEY, not(equalTo(BODY_KEY)))
    }

    @Test
    fun getPathParams() {
        val params = PathParams(emptyMap())
        doReturn(params).`when`(context).get<Any?>(PATH_PARAMS_KEY)
        assertThat(getPathParams(context), equalTo(params))
    }

    @Test
    fun getPathParamsWhenNoParams() {
        expect { getPathParams(context) }.toThrow<IllegalStateException>()
    }

    @Test
    fun putPathParams() {
        val params = PathParams(emptyMap())
        putPathParams(context, params)
        verify(context).put(PATH_PARAMS_KEY, params)
    }

    @Test
    fun getQueryParams() {
        val params = mock<QueryParams>()
        doReturn(params).`when`(context).get<Any?>(QUERY_PARAMS_KEY)
        assertThat(getQueryParams(context), equalTo(params))
    }

    @Test
    fun getQueryParamsWhenNoParams() {
        expect { getQueryParams(context) }.toThrow<IllegalStateException>()
    }

    @Test
    fun putQueryParams() {
        val params = mock<QueryParams>()
        putQueryParams(context, params)
        verify(context).put(QUERY_PARAMS_KEY, params)
    }

    @Test
    fun getDecodedBody() {
        doReturn("expected").`when`(context).get<Any?>(BODY_KEY)
        assertThat(getDecodedBody(context), equalTo("expected"))
    }

    @Test
    fun getOptionalDecodedBody() {
        doReturn("expected").`when`(context).get<Any?>(BODY_KEY)
        assertThat(getOptionalDecodedBody(context), equalTo("expected"))
    }

    @Test
    fun getMissingOptionalDecodedBody() {
        assertThat(getOptionalDecodedBody(context), nullValue())
    }

    @Test
    fun geRequestBodyWhenNoBody() {
        expect { getDecodedBody(context) }.toThrow<BadParamException>()
    }

    @Test
    fun putRequestBody() {
        val body = Any()
        putRequestBody(context, body)
        verify(context).put(BODY_KEY, body)
    }

}
